### PR TITLE
File upload service

### PR DESCRIPTION
**Pull request by writemorecode**
This PR implements a simple file storage service.

Operations:

- Upload files (`POST /files`, supports bulk uploads).
- Delete file and metadata (`DELETE /files/{file_id}`)
- Get file metadata (`GET /files/{file_id}`)
- Get all metadata for user (`GET /files`)
- Download file (`GET /files/{file_id}`)

The filesystem path used to store files is configured using the `FILES_UPLOAD_PATH` environment variable in `env.local`. For security reasons, files are not stored on disk under their original filenames. Instead, they are named as the ID field of the record ID of the metadata record in the database.

### CODE REVIEW

File: schemas/files_for.surql [hunk 1]
```diff
@@ -0,0 +1,2 @@
+DEFINE FIELD IF NOT EXISTS in ON files_for TYPE record<file> PERMISSIONS FULL;
```
Comment (retrokiller543) [inline schemas/files_for.surql:1]: Update the query to use the built in way to create relation tables https://surrealdb.com/docs/surrealql/statements/define/table

File: src/endpoints/api/files/mod.rs [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+use utoipa::path;
+use utoipa::OpenApi;
+
+pub struct FilesServiceState {
+    pub upload_path: PathBuf,
+}
+
+impl FilesServiceState {
+    pub fn new() -> Self {
+        let env_upload_path = tosic_utils::prelude::env!("FILES_UPLOAD_PATH");
```
Comment (retrokiller543) [inline src/endpoints/api/files/mod.rs:33]: would maybe not say that this is a required env var, we could fallback to current dir or use the crate `dirs` to fetch the tmp dir and store them there

File: src/endpoints/api/files/mod.rs [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+use std::fs;
+use std::path::PathBuf;
+use std::sync::Arc;
+use surrealdb::sql::Thing;
+use surrealdb::Surreal;
+use utoipa::openapi;
+use utoipa::path;
+use utoipa::OpenApi;
+
+pub struct FilesServiceState {
```
Comment (retrokiller543) [inline src/endpoints/api/files/mod.rs:27]: Maybe move state out of the endpoint module

File: src/endpoints/api/files/mod.rs [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+        Self { upload_path }
+    }
+
+    pub fn get_path_for(&self, filename: &str) -> PathBuf {
+        self.upload_path.join(filename)
+    }
+}
+
+#[derive(Debug, MultipartForm)]
+struct UploadForm {
```
Comment (retrokiller543) [inline src/endpoints/api/files/mod.rs:44]: Move DTO to dto module

File: src/endpoints/api/files/mod.rs [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+    tag = "files",
+    responses (
+        (status = 200, response = FileMetadata),
+        (status = 401, description = "Unauthorized"),
+        (status = 404, description = "File not found"),
+        (status = 500, description = "Internal server error"),
+    ),
+)]
+#[get("/{file_id}")]
+async fn get_file(
```
Comment (retrokiller543) [inline src/endpoints/api/files/mod.rs:103]: consider using the generate_endpoint macro and possibly split the endpoints into several sepperate fiels, one for each endpoint/http method. Moving it to sepparete files is optional but could make finding in the project easier

File: src/endpoints/api/files/mod.rs [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+    }
+}
+
+#[derive(Debug, MultipartForm)]
+struct UploadForm {
+    #[multipart(rename = "file")]
+    files: Vec<TempFile>,
+}
+
+async fn token_from_request<T>(
```
Comment (retrokiller543) [inline src/endpoints/api/files/mod.rs:49]: move this to the extractors module where other extraction logic lives

File: src/models/file_metadata.rs [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+    /// For security reasons, uploaded files are
+    /// persisted to disk with a filename equal
+    /// to its database record ID.
+    pub filename: String,
+    /// The datetime when the file was uploaded.
+    pub created_at: Datetime,
+}
+
+#[derive(ToResponse)]
+pub struct FileMetadataMultiple {
```
Comment (retrokiller543) [inline src/models/file_metadata.rs:19]: consider a tuple struct instead, we dont really need the named field

File: src/services/files.rs [hunk 8]
```diff
@@ -0,0 +1,113 @@
+use crate::error::ServerResponseError;
+use crate::models::file_metadata::FileMetadata;
+use crate::services::user::get::get_user_by_token;
+use serde::{Deserialize, Serialize};
+use std::sync::Arc;
+use surrealdb::sql::Thing;
+use surrealdb::Surreal;
+
+#[derive(Debug, Serialize, Deserialize)]
+struct Filename {
```
Comment (retrokiller543) [inline src/services/files.rs:10]: consider tuple struct unless this is stricly needed for Serialization or deserialization

File: src/services/files.rs [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+    filename: String,
+}
+
+impl Filename {
+    fn new(filename: &str) -> Self {
```
Comment (retrokiller543) [inline src/services/files.rs:15]: if its a tuple struct there is no need for a new method we could then construct this by using `Filename(filename)`

File: src/services/files.rs [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+        Self {
+            filename: filename.to_string(),
+        }
+    }
+}
+
+/// Inserts metadata for multiple files and relates them
+/// to the user holding token `token`.
+pub async fn insert_file_metadata<T>(
```
Comment (retrokiller543) [inline src/services/files.rs:24]: split logic into sepperate files, i would prefer if the structure is something like this

```
files
> mod.rs
> insert.rs
> get.rs
> delete.rs
```

and then re-export them in the `mod.rs`

File: src/services/files.rs [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+        .bind(("USER", user_id))
+        .await?
+        .take(2)?;
+
+    Ok(created)
+}
+
+/// Returns the metadata of the file with ID `file_id` uploaded by
+/// the user holding the token `token`.
+pub async fn get_file_metadata<T>(
```
Comment (retrokiller543) [inline src/services/files.rs:60]: split logic into sepperate files, i would prefer if the structure is something like this

```
files
> mod.rs
> insert.rs
> get.rs
> delete.rs
```

and then re-export them in the `mod.rs`

File: src/services/files.rs [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+        .take(0)?;
+    match found {
+        Some(file) => Ok(file),
+        None => Err(ServerResponseError::NotFound),
+    }
+}
+
+/// Returns metadata of all files uploaded by the user
+/// holding token `token`.
+pub async fn get_file_metadata_by_token<T>(
```
Comment (retrokiller543) [inline src/services/files.rs:84]: split logic into sepperate files, i would prefer if the structure is something like this

```
files
> mod.rs
> insert.rs
> get.rs
> delete.rs
```

and then re-export them in the `mod.rs`

File: src/services/files.rs [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+    T: surrealdb::Connection,
+{
+    const SQL: &str = "SELECT VALUE in FROM files_for WHERE out = $USER FETCH in;";
+    let files: Vec<FileMetadata> = db.query(SQL).bind(("USER", user_id)).await?.take(0)?;
+    Ok(files)
+}
+
+/// Deletes the metadata of a file with ID `file_id` that
+/// was uploaded by the user holding the token `token`.
+pub async fn delete_file_metadata<T>(
```
Comment (retrokiller543) [inline src/services/files.rs:98]: split logic into sepperate files, i would prefer if the structure is something like this

```
files
> mod.rs
> insert.rs
> get.rs
> delete.rs
```

and then re-export them in the `mod.rs`

File: src/state.rs [hunk 14]
```diff
@@ -9,6 +10,7 @@ use surrealdb::Surreal;
 pub struct AppState {
     pub db: Arc<Surreal<Client>>,
     pub oauth: Arc<Oauth>,
+    pub files: FilesServiceState,
```
Comment (retrokiller543) [inline src/state.rs:13]: One idea i have had is that we could split the state objects into their own `web::Data` items making it simpler to retrieve the desired state that we need from the API state and ignore any state that is not needed.

**Review by retrokiller543 — CHANGES_REQUESTED** [review]

File: src/services/files.rs (hunk 8 above)
Comment (writemorecode) [reply src/services/files.rs:10]: I tried that, doesn't work. The "filename" field name is used for deserialization to build the query.

File: src/endpoints/api/files/mod.rs (hunk 3 above)
Comment (retrokiller543) [reply src/endpoints/api/files/mod.rs:29]: this is still not resolved, please fix before we merge

File: src/endpoints/api/files/mod.rs (hunk 2 above)
Comment (retrokiller543) [reply src/endpoints/api/files/mod.rs:35]: this is still not resolved, please fix before we merge

File: src/endpoints/api/files/mod.rs (hunk 5 above)
Comment (retrokiller543) [reply src/endpoints/api/files/mod.rs:57]: Why was this resolved? i cant see a reason for it

**Review by retrokiller543 — CHANGES_REQUESTED** [review]
Most comments was resolved but not all and one was resolved with no fix or comment about why it was not resolved

File: src/endpoints/api/files/mod.rs (hunk 2 above)
Comment (writemorecode) [reply src/endpoints/api/files/mod.rs]: I've now fixed this.

File: src/endpoints/api/files/mod.rs (hunk 3 above)
Comment (writemorecode) [reply src/endpoints/api/files/mod.rs]: I've now fixed this.

**Review by retrokiller543 — APPROVED** [review]

**Pull request merged**